core/distributed/matrix.cpp

```diff
@@ -369,15 +371,13 @@ void Matrix<ValueType, LocalIndexType, GlobalIndexType>::apply_impl(
             local_mtx_->apply(dense_b->get_local_vector(), local_x);
             req.wait();
 
-            if (non_local_mtx_) {
-                auto exec = this->get_executor();
-                auto use_host_buffer = mpi::requires_host_buffer(exec, comm);
-                if (use_host_buffer) {
-                    recv_buffer_->copy_from(host_recv_buffer_.get());
-                }
-                non_local_mtx_->apply(one_scalar_.get(), recv_buffer_.get(),
-                                      one_scalar_.get(), local_x);
+            auto exec = this->get_executor();
+            auto use_host_buffer = mpi::requires_host_buffer(exec, comm);
+            if (use_host_buffer) {
+                recv_buffer_->copy_from(host_recv_buffer_.get());
```

Comment:
Now you need to do an extra copy (when using host_buffers) even if there is no need to call `non_local_mtx->apply()`, right ? 

IMO extra `if` does not complicate the control flow a lot.

Reply:
If non-local is empty, then so is the size of the recv_buffer (excluding the pathological case of size `0 x N`). So all operations within the if are no-ops anyway.